api/crossmodelrelations/macarooncache_test.go

```diff
@@ -0,0 +1,120 @@
+
+import (
+	"time"
+
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/macaroon.v1"
+
+	"github.com/juju/juju/api/crossmodelrelations"
+	coretesting "github.com/juju/juju/testing"
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+)
+
+var _ = gc.Suite(&MacaroonCacheSuite{})
+
+type MacaroonCacheSuite struct {
+	coretesting.BaseSuite
+}
+
+func (s *MacaroonCacheSuite) TestGetMacaroonMissing(c *gc.C) {
+	cache := crossmodelrelations.NewMacaroonCache(testing.NewClock(time.Now()))
+	_, ok := cache.Get("missing")
+	c.Assert(ok, jc.IsFalse)
+}
+
+func (s *MacaroonCacheSuite) TestGetMacaroon(c *gc.C) {
+	cache := crossmodelrelations.NewMacaroonCache(testing.NewClock(time.Now()))
+	mac, err := macaroon.New(nil, "", "")
+	c.Assert(err, jc.ErrorIsNil)
+	cache.Upsert("token", macaroon.Slice{mac})
+	ms, ok := cache.Get("token")
+	c.Assert(ok, jc.IsTrue)
+	c.Assert(ms, jc.DeepEquals, macaroon.Slice{mac})
+}
+
+func (s *MacaroonCacheSuite) TestGetMacaroonNotExpired(c *gc.C) {
+	clock := testing.NewClock(time.Now())
+	cache := crossmodelrelations.NewMacaroonCache(clock)
+
+	mac, err := macaroon.New(nil, "", "")
+	cav := checkers.TimeBeforeCaveat(clock.Now().Add(10 * time.Second))
+	mac.AddFirstPartyCaveat(cav.Condition)
+	c.Assert(err, jc.ErrorIsNil)
+
+	cache.Upsert("token", macaroon.Slice{mac})
+	clock.Advance(9 * time.Second)
+
+	ms, ok := cache.Get("token")
+	c.Assert(ok, jc.IsTrue)
+	c.Assert(ms, jc.DeepEquals, macaroon.Slice{mac})
+}
+
+func (s *MacaroonCacheSuite) TestGetMacaroonExpiredBeforeCleanup(c *gc.C) {
+	clock := testing.NewClock(time.Now())
+	cache := crossmodelrelations.NewMacaroonCache(clock)
+
+	mac, err := macaroon.New(nil, "", "")
+	cav := checkers.TimeBeforeCaveat(clock.Now().Add(10 * time.Second))
+	mac.AddFirstPartyCaveat(cav.Condition)
+	c.Assert(err, jc.ErrorIsNil)
+
+	cache.Upsert("token", macaroon.Slice{mac})
+	clock.Advance(20 * time.Second)
+
+	_, ok := cache.Get("token")
+	c.Assert(ok, jc.IsFalse)
+}
+
+func (s *MacaroonCacheSuite) TestGetMacaroonAfterCleanup(c *gc.C) {
+	clock := testing.NewClock(time.Now())
+	cache := crossmodelrelations.NewMacaroonCache(clock)
+
+	mac, err := macaroon.New(nil, "", "")
+	cav := checkers.TimeBeforeCaveat(clock.Now().Add(10 * time.Minute))
+	mac.AddFirstPartyCaveat(cav.Condition)
+	c.Assert(err, jc.ErrorIsNil)
+
+	cache.Upsert("token", macaroon.Slice{mac})
+	clock.Advance(6 * time.Minute)
+
+	ms, ok := cache.Get("token")
+	c.Assert(ok, jc.IsTrue)
+	c.Assert(ms, jc.DeepEquals, macaroon.Slice{mac})
+}
+
+func (s *MacaroonCacheSuite) TestMacaroonRemovedByCleanup(c *gc.C) {
+	clock := testing.NewClock(time.Now())
+	cache := crossmodelrelations.NewMacaroonCache(clock)
+
+	mac, err := macaroon.New(nil, "", "")
+	cav := checkers.TimeBeforeCaveat(clock.Now().Add(2 * time.Minute))
+	mac.AddFirstPartyCaveat(cav.Condition)
+	c.Assert(err, jc.ErrorIsNil)
+
+	cache.Upsert("token", macaroon.Slice{mac})
+	clock.Advance(6 * time.Minute)
```

Comment:
Sometimes this won't test what it thinks it's testing - if the expiry goroutine hasn't gotten to the `clock.After` call by the time the test calls `Advance` it won't see the update. It won't be unreliable since the expiry check in `Get` will still expire it anyway, but this should probably still use `WaitAdvance`.